Add `intersects` function

seadowg opened the pull request:
Closes #6963

This adds the function so it can be used in constraints for `geoshape` and `geotrace` questions. There's one deviation from the issue:

> If the parameter type is not geotrace or geoshape, an error is thrown

Instead of throwing an error, the function will just return `false` in this scenario. We can discuss in the issue if that's correct or not, and if I'm wrong follow-up with another PR.

#### Why is this the best possible solution? Were any other approaches considered?

I chose to add this using JavaRosa's [Function plugin system](https://github.com/getodk/javarosa/blob/master/PLUGINS.md#function) rather than making changes in JavaRosa itself. I'm very in favour for using JavaRosa's plugins as much as possible: it means we limit changes to and the size of JavaRosa itself, allows us to write code in Kotlin, and we don't have to do a JavaRosa release to add a new feature. We could consider moving the other geo specific functions (`area`, `distance` and `geofence`) to Collect as well - I'll let @grzesiek2010 tell me if he thinks I should follow up with that or not.

In terms of the function itself, I had to relearn a bunch of geometry of various levels I'd forgotten to work out how to implement intersection detection as it didn't seem like there was a widely available library I could pull in (outside something like AWT). Checking my thinking here is probably the most important part of review as this is the kind of thing where automated testing can only give us so much confidence.

The solution I ended up with uses what seems like a fairly standard trick in computational geometry where you can determine the "orientation" (or "direction") of a set of points by [calculating their cross product](https://en.wikipedia.org/wiki/Cross_product#Computational_geometry). You can then use this orientation to work out whether the endpoints of a line segment are on the either side of another line. There are definitely other solutions to this problem: you can solve it algebraically using linear equations and there's also [specific algorithms for finding self intersections in polygons](https://en.wikipedia.org/wiki/Bentley%E2%80%93Ottmann_algorithm). I wanted to start with something that was hopefully as easy to understand as possible though to someone jumping into the code fresh or after some time. I imagine there are definitely things we can do to optimize even just the approach we have so far (which currently has a worst case of `On^2` with respect to checking line segments against others) before looking at more complex solutions, but let's make sure we need to do that before going down that route.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

As I mentioned above, the biggest risk here is that I've messed up the maths of checking how lines intersect so trying to throw both realistic and weird examples at this would be great. I also think trying this on larger traces/shapes would be interesting to get a sense of performance. Other than `intersects` though, there should be no changes to the app.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] added any new strings with date formatting to `DateFormatsTest`
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

seadowg committed:
Add intersects util

seadowg committed:
Add intersects function handler

seadowg committed:
Use prototypes to get already evaluated args

seadowg committed:
Add test for non geotrace/shape strings

seadowg committed:
Add test cases for repeated segments and closed traced intersection

seadowg committed:
Add intersects function handler to Collect

seadowg commented on collect_app/src/main/java/org/odk/collect/android/formmanagement/CollectFormEntryControllerFactory.kt:
```diff
@@ -46,6 +47,9 @@ class CollectFormEntryControllerFactory(
                 )
             )
+
+            it.addFunctionHandler(IntersectsFunctionHandler())
```
I'm in two minds about this being untested. It can be seen as pure "configuration" from one angle, but also if I delete it then `intersects` won't work. Maybe it's worth replacing one of the `IntersectsFunctionHandler` tests (probably the detects intersection path) with a feature test?

grzesiek2010 replied:
> Maybe it's worth replacing one of the IntersectsFunctionHandler tests (probably the detects intersection path) with a feature test?

I think it would make sense.

seadowg committed:
Move parseGeometry and parseGeometryPoint tests to geo

seadowg committed:
Convert GeoUtils to Kotlin

seadowg committed:
Move parseGeometryPoint to GeoUtils

seadowg commented on geo/src/main/java/org/odk/collect/geo/javarosa/IntersectsFunctionHandler.kt:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    override fun realTime(): Boolean {
+        TODO("Not yet implemented")
```
I'm having a separate conversation with @lognaturel about the history of this method as it's not called by JavaRosa currently. Will follow-up with an update later.

grzesiek2010 replied:
So should we keep TODO for now?

seadowg replied:
Yeah for now.

seadowg committed:
Fix import in test

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/GeoUtils.kt:
```diff
@@ -0,0 +1,84 @@
+
+    @JvmStatic
+    fun parseGeometryPoint(answer: String?): DoubleArray? {
+        if (answer != null && answer.isNotEmpty()) {
```
This could be just `!answer.isNullOrBlank()`

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/GeoUtils.kt:
```diff
@@ -0,0 +1,84 @@
+    @JvmStatic
+    fun parseGeometryPoint(answer: String?): DoubleArray? {
+        if (answer != null && answer.isNotEmpty()) {
+            val sa = answer.trim { it <= ' ' }.split(" ").toTypedArray()
```
`toTypedArray()` is redundant.

grzesiek2010 replied:
It can be just `trim()`. We don't need to pass `{ it <= ' ' }` right?

seadowg replied:
> It can be just trim(). We don't need to pass { it <= ' ' } right?

Interestingly, Java and Kotlin's `trim` are slightly different in terms of how many different whitespace characters they cover - I think this is the Kotlin converter trying to make sure you don't end up clearing whitespace characters you wanted to keep. I don't think there's any chance of that here (as we're just dealing with geo strings), so we can safely use Kotlin's `trim`.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/GeoUtils.kt:
```diff
@@ -0,0 +1,84 @@
+                    if (sa.size > 2) sa[2].toDouble() else 0.0,
+                    if (sa.size > 3) sa[3].toDouble() else 0.0
+                )
+            } catch (e: Throwable) {
```
`e` is never used, so it should be replaced with `_`.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/geopoly/GeoPolyUtils.kt:
```diff
@@ -0,0 +1,109 @@
+    fun parseGeometry(geometry: String?): ArrayList<MapPoint> {
+        val points = ArrayList<MapPoint>()
+
+        for (vertex in (geometry ?: "").split(";").toTypedArray()) {
```
`toTypedArray()` is redundant.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/geopoly/GeoPolyUtils.kt:
```diff
@@ -0,0 +1,109 @@
+
+object GeoPolyUtils {
+
+    fun parseGeometry(geometry: String?): ArrayList<MapPoint> {
```
The method should use the interface in its declaration (List).

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/GeoUtils.kt:
```diff
@@ -0,0 +1,84 @@
+            )
+        }
+
+        return removeEnd(result.toString().trim { it <= ' ' }, ";")
```
It can be just `trim()`. We don't need to pass `{ it <= ' ' }` right?

grzesiek2010 commented on geo/src/test/java/org/odk/collect/geo/GeoUtilsTest.kt:
```diff
@@ -59,4 +61,33 @@ class GeoUtilsTest {
+    fun parseGeometryPointTest() {
+        var gp =
+            parseGeometryPoint("37.45153333333334 -122.15539166666667 0.0 20.0")!!
+        TestCase.assertEquals(37.45153333333334, gp[0])
```
Let's use `assertThat` like in other tests.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/geopoly/GeoPolyUtils.kt:
```diff
@@ -0,0 +1,109 @@
+    fun intersects(trace: List<MapPoint>): Boolean {
+        return if (trace.size >= 3) {
+            val segments = getSegments(trace)
+            segments.any { line1 ->
```
We don't need to compare segments with themselves or their neighbors, so this can be optimized for better performance.

seadowg replied:
From the PR description:

> I imagine there are definitely things we can do to optimize even just the approach we have so far (which currently has a worst case of On^2 with respect to checking line segments against others) before looking at more complex solutions, but let's make sure we need to do that before going down that route.

I definitely agree that even just this `On^2` loop can be improved (skipping self checks etc), but I think we want to get a sense of what level of improvements we need to make and add some form of performance testing or benchmarking to drive them out before trying to optimize.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/geopoly/GeoPolyUtils.kt:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    private fun getSegments(trace: List<MapPoint>): List<Pair<MapPoint, MapPoint>> {
+        return trace.flatMapIndexed { index, point ->
```
You can simply use `zipWithNext()`.

seadowg commented:
@grzesiek2010 marking this as a draft while I add the error cases discussed in https://github.com/getodk/collect/issues/6963. It won't have any significant impact, so feel free to keep reviewing if you already started.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/javarosa/IntersectsFunctionHandler.kt:
```diff
@@ -0,0 +1,32 @@
+        args: Array<out Any?>,
+        ec: EvaluationContext
+    ): Any {
+        val trace = parseGeometry(args[0] as String)
```
Should we throw an error if the number of args is different than 1?

grzesiek2010 replied:
Ah there will be an error because rawArgs returns false. this should be tested in IntersectsFunctionHandlerTest.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/javarosa/IntersectsFunctionHandler.kt:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    override fun eval(
+        args: Array<out Any?>,
```
It doesn't need to be nullable?

seadowg replied:
I'll clean the signature of this for Kotlin up as part of updating docs and working out what to do with `realTime`.

seadowg committed:
Throw exception for non-geotrace inputs

seadowg committed:
Improve converted code

seadowg committed:
Use zipWithNext to simplify creating segments

seadowg committed:
Add test to check arg length is enforced

seadowg committed:
Switch test to Hamcrest

grzesiek2010 commented on geo/src/test/java/org/odk/collect/geo/GeoUtilsTest.kt:
```diff
@@ -25,30 +26,30 @@ class GeoUtilsTest {
 
     @Test
     fun whenPointsAreNull_formatPoints_returnsEmptyString() {
-        assertEquals(formatPointsResultString(emptyList(), true), "")
```
The import is not removed and it is unused. The build should fail...

seadowg replied:
Ok so it looks like `no-unused-imports` isn't run by ktlint any more by default: https://github.com/pinterest/ktlint/blob/master/CHANGELOG.md#170---2025-07-14. I've re-enabled it for now, but we may need to find another solution if they do go ahead and remove it in 2.0.

seadowg committed:
Remove unused imports

seadowg committed:
Add another test case to make sure we check intersections from both segments

grzesiek2010 commented on geo/src/test/java/org/odk/collect/geo/geopoly/GeoPolyUtilsTest.kt:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Test
+    fun `#intersects returns false when a segment's end points are both on different sides of another, but the segments do not intersect`() {
```
Just an idea, if you don't like it, please ignore...
It would be easier to work with those tests if visualizing were easier. Maybe we could add comments like:
```
    /*
     *  3 | D   .   .   C
     *  2 | .   B   .   .
     *  1 | .   A   .   .
     *  0 | .   .   .   .
     *    +----------------
     *      0   1   2   3
     */
```

seadowg replied:
I'd been struggling to think of a nice way to represent these! Did you find a way to generate that or is it by hand?

grzesiek2010 replied:
It was ChatGPT, but I had to correct the output.

seadowg commented:
@grzesiek2010 I'm going to mark this as a draft again while I deal with a couple of cases I don't think our current algorithm covers. There won't be any structural changes.

seadowg committed:
Add failing tests for non crossing intersections

seadowg committed:
Rename method

seadowg committed:
Fix origin of one segment touching another case

seadowg committed:
Correct test

The example before would be considered closed and so wouldn't
be classes as intersecting

seadowg committed:
Fix line moving back on itself case

seadowg committed:
Add test for shape that closes outside the origin

seadowg commented on geo/src/main/java/org/odk/collect/geo/geopoly/GeoPolyUtils.kt:
```diff
@@ -0,0 +1,130 @@
+            segments.filterIndexed { line1Index, line1 ->
+                segments.filterIndexed { line2Index, line2 ->
+                    if (line2Index >= line1Index + 2) {
+                        crosses(line1, line2) || within(line1.first, line2)
```
I think this should probably also do `within(line1.second, line2)` and also check line 2's endpoints vs line1 to truly check for intersection, but I can't think of a shape/trace that would need that check. I think the nature of all the segments following each other and connecting and vertices prevents other situations. @grzesiek2010 any ideas of things that might break?

It might be a good tactic to just pull out a `intersects` function that works on two line segments to use here. Then we could just test that more exhaustively for intersections that aren't possible within a trace. That's maybe more than we need, but give us more confidence that we're not missing anything.

grzesiek2010 replied:
> I think this should probably also do within(line1.second, line2) and also check line 2's endpoints vs line1 to truly check for intersection, but I can't think of a shape/trace that would need that check. 

I think we don't need that. At least I can't think of a ces that would break it.

grzesiek2010 commented on geo/src/test/java/org/odk/collect/geo/geopoly/GeoPolyUtilsTest.kt:
```diff
@@ -0,0 +1,142 @@
+    fun `#intersects returns true when the trace closes on a non-origin vertex`() {
+        val trace = listOf(
+            MapPoint(0.0, 0.0),
+            MapPoint(0.0, 1.0),
```
This point could be skipped, I think.

seadowg replied:
The idea is to close on a non-origin vertex which is this one. I can add a comment, and there probably is a smaller example.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/geopoly/GeoPolyUtils.kt:
```diff
@@ -0,0 +1,130 @@
+    /**
+     * Returns `true` if any segment of the trace intersects with any other and `false` otherwise.
+     */
+    fun intersects(trace: List<MapPoint>): Boolean {
```
I played with two cases:

```
    @Test
    fun `test`() {
        val trace = listOf(
            MapPoint(0.0, 0.0),
            MapPoint(5.0, 0.0),
            MapPoint(0.0, 0.0)
        )

        assertThat(GeoPolyUtils.intersects(trace), equalTo(false))
    }
```
passes but

```
    @Test
    fun `test`() {
        val trace = listOf(
            MapPoint(0.0, 0.0),
            MapPoint(5.0, 0.0),
            MapPoint(4.0, 0.0),
            MapPoint(0.0, 0.0)
        )

        assertThat(GeoPolyUtils.intersects(trace), equalTo(false))
    }
```

fails.

In fact, the difference is just the number of points but on the map they would look the same so the result should be the same as well.

seadowg replied:
Both of these should return `true` right? The first one is a special case where there the shape is closed, there are two segments and the segments are just the reverse of each other and this fools our code because we don't check the last point for intersection on a closed trace. I think we might need to capture it specifically.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/geopoly/GeoPolyUtils.kt:
```diff
@@ -0,0 +1,130 @@
+    /**
+     * Check if a point is within a line
+     */
+    fun within(
```
The current implementation of `within()` is incorrect because it only checks if the point is inside the bounding box of the segment, without verifying collinearity. This can lead to false positives where a point appears to be on the segment but is actually far from the line.

A test case illustrating the issue:

```
val trace = listOf(
        MapPoint(0.0, 0.0),
        MapPoint(10.0, 10.0),
        MapPoint(0.0, 10.0)
)
GeoPolyUtils.intersects(trace)  // should return false
```

seadowg committed:
Simplify non-origin vertex closing

seadowg committed:
Use bounding box check based on orientation to solve right angled triangle case

seadowg committed:
Fix special case with two segmnt intersection

seadowg committed:
Pull out specific 2D geometry code

seadowg committed:
Make sure checking instersection between segments is exhaustive

seadowg commented:
@grzesiek2010 thanks so much for helping with examples. I was pretty sure that we needed additional checks in the collinear endpoints scenarios (I'd seen it be exhaustive in examples), but was really struggling with finding tests to prove that. I think we should have a proper intersects check now!

grzesiek2010 commented on shared/src/main/java/org/odk/collect/shared/geometry/Geometry.kt:
```diff
@@ -0,0 +1,130 @@
+        val segments = segments()
+        return if (segments.size == 2) {
+            val orientation = orientation(segments[1].end, segments[0].start, segments[0].end)
+            return orientation == Orientation.Collinear && segments[1].end.within(segments[0])
```
You don't need these two returns (here and in line 25).

grzesiek2010 commented on shared/src/main/java/org/odk/collect/shared/geometry/Geometry.kt:
```diff
@@ -0,0 +1,130 @@
+/**
+ * Returns `true` if any segment of the trace intersects with any other and `false` otherwise.
+ */
+fun Trace.intersects(): Boolean {
```
The way we collect points probably already handles cases where points are duplicated, but this code doesn’t need to know about that.
As a result, this test fails:

```
val trace = Trace(
            listOf(
                Point(0.0, 0.0),
                Point(1.0, 0.0),
                Point(1.0, 0.0),
                Point(4.0, 0.0),
            )
        )
        assertThat(trace.intersects(), equalTo(false))
```

but I don’t think it should.

seadowg replied:
I'm actually not so sure! Here you have a trace where the segments are like this:

1. 0,0 -> 1,0
2. 1,0 -> 1,0
3. 1,0 -> 4,0

So because of the duplicate points, we end up with a 0 length segment and that's detected because segments 1 and 3 share a point (1,0). This is allowed for consecutive segments, but not for non-consecutive segments. I think this probably should be counted as intersection. I'll add a discussion item to our sync so we can clarify.

grzesiek2010 commented on shared/src/main/java/org/odk/collect/shared/geometry/Geometry.kt:
```diff
@@ -0,0 +1,130 @@
+    } else if (orientationC == Orientation.Collinear && c.within(other)) {
+        true
+    } else if (orientationD == Orientation.Collinear && d.within(other)) {
+        true
```
`LineSegment.intersects(`) currently uses a collinear + bounding-box check (point.within(other)), which can produce false positives. In these tests, some segments are collinear and a segment endpoint lies within the bounding box of another segment, but the segments do not actually intersect:
```
        val trace = Trace(
            listOf(
                Point(0.0, 0.0), 
                Point(4.0, 0.0),
                Point(2.0, 2.0),
                Point(4.0, 4.0), 
                Point(0.0, 4.0),
                Point(0.0, 0.0)  
            )
        )

        assertThat(trace.intersects(), equalTo(false))
```

grzesiek2010 replied:
This as well:
```
       val trace = Trace(
            listOf(
                Point(0.0, 0.0),
                Point(4.0, 0.0),
                Point(4.0, 4.0),
                Point(8.0, 0.0)
            )
        )
        assertThat(trace.intersects(), equalTo(false))
```

seadowg replied:
Ah good catch! I don't think it's a mistake in the approach, but it is a mistake in the implementation: the bounding box check for points that have collinear orientation vs `this` are being checked against `other` which is the line they are on. For `c` and `d` it should be `c.within(this)`. Let me know if you disagree.

seadowg committed:
Remove unneeded return

seadowg committed:
Correct bounding box check direction

grzesiek2010 commented on shared/src/main/java/org/odk/collect/shared/geometry/Geometry.kt:
```diff
@@ -0,0 +1,130 @@
+/**
+ * Returns `true` if any segment of the trace intersects with any other and `false` otherwise.
+ */
+fun Trace.intersects(): Boolean {
```
This test fails here:
```
    @Test
    fun `test 3 points with first point on last segment`() {
        val trace = Trace(listOf(
            Point(1.0, 0.0),
            Point(0.0, 0.0),
            Point(2.0, 0.0),
        ))
        assertThat(trace.intersects(), equalTo(true))
    }
```

grzesiek2010 commented on shared/src/main/java/org/odk/collect/shared/geometry/Geometry.kt:
```diff
@@ -0,0 +1,130 @@
+    val points = this.points
+    return if (points.size >= 3) {
+        val segments = segments()
+        return if (segments.size == 2) {
```
This return can be removed as well.

seadowg committed:
Add additional (failing) test for 2 segment self intersection

seadowg committed:
Account for case where first endpoint intersects with second segment in 2 segment connected case

seadowg committed:
Filter out zero length segments in traces

seadowg committed:
Remove unneeded return

grzesiek2010 commented:
One more test that fails:

```
    @Test
    fun `opposite direction segments overlapping`() {
        val trace = Trace(listOf(
            Point(0.0, 0.0),
            Point(2.0, 2.0),
            Point(1.0, 1.0),
            Point(0.0, 0.0)
        ))
        assertThat(trace.intersects(), equalTo(true))
    }
```

grzesiek2010 reviewed (changes requested):
https://github.com/getodk/collect/pull/6967#issuecomment-3583270669

grzesiek2010 commented:
@seadowg I’ve found one more case that needs to be handled. I don’t think I’ll be able to find more without spending hours on it. However, after discovering so many cases, I’m not fully confident in accepting the implementation as is. I think this is a perfect case for AI. Maybe we could ask it to generate even 1,000 test cases to cover different scenarios and ensure we don’t miss anything. What do you think?

seadowg commented:
> I don’t think I’ll be able to find more without spending hours on it. However, after discovering so many cases, I’m not fully confident in accepting the implementation as is.

Agreed! I think we've probably hit the limit of how much confidence we can get from this "ping-pong pairing" style back and forth to generate test cases for ourselves. Beyond this, it'll get harder to think of, describe and later understand cases we might want to test.

> I think this is a perfect case for AI. Maybe we could ask it to generate even 1,000 test cases to cover different scenarios and ensure we don’t miss anything. What do you think?

I do think using some form of automated test case generation is needed here. Using an LLM to generate cases could work, but it would add a lot of noise to our test code, and we'd either need to validate each one or accept that there is a risk of false positives from LLM misunderstanding or hallucination.

However, there are some non-AI approaches we could use here. My mind immediately jumps to using something like [QuickCheck](https://en.wikipedia.org/wiki/QuickCheck) where we generate random inputs (here traces) and then just test that our function behaves correctly with them. An example of this would be testing an `isEven` helper:

```kotlin
// This function gets passed a random number as many times as we think we need
fun testIsEven(value: Int) {
    if (value % 2 == 0) {
        assertThat(value.isEven(), equalTo(true))
    } else {
        assertThat(value.isEven(), equalTo(false))
    }
}
```

The problem we have with `intersects` is that we can't verify that our function behaves correctly without using an existing implementation. I looked this up, and apparently that's reffered to as the ["test oracle problem"](https://en.wikipedia.org/wiki/Test_oracle). As an aside, this also make using an LLM to generate test cases hard: we can't validate that the LLM is producing correct tests without already having a correct implementation.

As mentioned in the "test oracle" wikipedia, one approach I have a vague memory of learning about, but haven't really used is (the very academically named) ["metamorphic testing"](https://en.wikipedia.org/wiki/Metamorphic_testing). Here you use the QuickCheck approach, but instead of testing the outputs directly, you test modifications to the input/output that you know should hold:

> As an example, when testing a booking website, a [web search](https://en.wikipedia.org/wiki/Web_search_engine) for accommodation in Sydney, Australia, returns 1,671 results; are the results of this search correct and complete? This is a test oracle problem. Based on a metamorphic relation, we may filter the price range or star rating and apply the search again; it should return a subset of the previous results. A violation of this expectation would similarly reveal a failure of the system.

I think this should be a good technique here. We can easily generate (closed and unclosed) random traces and then test these "metamorphic relations":

- The result of `intersects` should be consistent when reversing (changing the order of the points), rotating or moving the trace
- If `intersects` is `false`, adding a segment that intersects should make it `false`

Let me know if there are other ones you can think of. This is obviously only a "fuzzy" approach, but without actually writing a proof that will always be the case. I think these checks (and the generation code) should be fairly easy to write, so I can have a go today and see how it looks so we can discuss.

I'll first look at solving https://github.com/getodk/collect/pull/6967#issuecomment-3583270669 obviously (which I think is just a mistake in how we deal with closed traces), but after that I think we should hand over to QA just to get more people playing with it.

seadowg committed:
Fix reversed line with 3 points case

seadowg committed:
Add basic metamorphic test for intersects

seadowg committed:
Improve adding intersecting segment using interpolation

seadowg committed:
Add epsilon to colinearity check to prevent precision errors

seadowg committed:
Fix accidental map

seadowg commented on shared/src/test/java/org/odk/collect/shared/geometry/GeometryTest.kt:
```diff
@@ -0,0 +1,302 @@
+    }
+
+    @Test
+    fun `Trace#intersects satisfies metamorphic relationships`() {
```
@grzesiek2010 Here's the metamorphic test I've added. It currently checks that intersection is consistent when reversing/scaling traces and that adding an intersecting segment to a non-intersecting trace causes `intersects` to become true. We can tidy this up/formalize it a bit, but I wanted to get your thoughts first! Right now, I'm checking 1000 random traces, but this could of course be less or more depending on how much time/power we want to expend.

seadowg replied:
You'll also see that I had to add an epsilon to the collinearity check to handle precision errors and that actually got revealed because of this test which was pretty cool. I'm going to involve @getodk/testers now so we can get some more feedback on using this. It'll be interesting to see if our current calculations are actually **too** precise when playing around in the real world.

seadowg committed:
Add docs for interpolate

grzesiek2010 commented on shared/src/test/java/org/odk/collect/shared/geometry/GeometryTest.kt:
```diff
@@ -0,0 +1,302 @@
+            // Check adding an intersection makes intersects true
+            if (!intersects) {
+                val intersectPosition = Random.nextDouble(0.0, 1.0)
+                val intersectionPoint = trace.segments().first().interpolate(intersectPosition)
```
Do I understand correctly that if `intersectPosition` is 0, then `trace.segments().first().interpolate(intersectPosition)` returns the first point of the first segment, and as a result, it will form a closed polygon with no intersections?

seadowg replied:
Ah yes! I was meaning to make it use a random segment rather than the first segment (and avoid this). I'll change that.

grzesiek2010 commented on shared/src/main/java/org/odk/collect/shared/geometry/Geometry.kt:
```diff
@@ -0,0 +1,159 @@
+ * [Linear interpolation](https://en.wikipedia.org/wiki/Linear_interpolation). `0` will return
+ * [LineSegment.start] and `1` will return [LineSegment.end].
+ */
+fun LineSegment.interpolate(position: Double): Point {
```
We should have tests for this method.

seadowg committed:
Increase the number of possible random intersection points in test

seadowg committed:
Add tests for LineSegment#interpolate

seadowg committed:
Add quickCheck helper

seadowg commented on shared/src/test/java/org/odk/collect/shared/QuickCheck.kt:
```diff
@@ -0,0 +1,16 @@
+ * Simple helper to allow writing neater [QuickCheck](https://en.wikipedia.org/wiki/QuickCheck)
+ * style tests.
+ */
+fun <Input, Output> ((Input) -> Output).quickCheck(
```
I did have a quick look at using an existing framework (https://github.com/pholser/junit-quickcheck for example), but it felt like we didn't really need anything so heavy yet. Maybe if we end up writing more of these kinds of tests down the line!

dbemke commented:
Should we test this PR on all map engines? If yes, could you send us the apk and check if Google maps token work in it?

seadowg commented:
> Should we test this PR on all map engines? If yes, could you send us the apk and check if Google maps token work in it?

Map engines won't matter so you can just test with OSM. The actual geotrace/shape features are untouched: this just adds a function you can use for constraints/calculations.

dbemke commented:
I found a case in which I can see lines which don't intersect but there's constraint message in the question.
<img width="440" height="831" alt="invalisShapw" src="https://github.com/user-attachments/assets/d30c8f28-fa27-40c4-ba39-5cefe759faea" />

I've put the geometry in a form with default values.
This was taken with automatic location recording set to 1 second and "none" accuracy and I was walking.

**Steps to reproduce:**
1. Scan user "invalid polygons".
https://test.getodk.cloud/projects/629/app-users
2. Open geoshape geatrace defaults example forms.
3. Go to "test bug" question and try to swipe to the next question.
4. Open the map.

grzesiek2010 commented:
@dbemke I think for reproducing and debugging, it's important to copy and paste the exact points that construct the polygon.

dbemke commented:
> I think for reproducing and debugging, it's important to copy and paste the exact points that construct the polygon.

and this is what I did - copied it to default in the form mentioned above. The link was missing.

WKobus commented:
**Tested with success**

Verified on Android 16, 11, 10

**Verified cases:**
- Intersects in Geoshape and Geotrace questions
- Different recording modes
- Mixing intersects and geofence
- Performance with lot of points on map

dbemke commented:
**Tested with success**

Verified on Android 8.1, 10
